### PR TITLE
ENH: find and replace functionality

### PR DESCRIPTION
**Pull request by tangkong**
## Description
- Implements atef-specific find and replace functionality
- adds a basic find-and-replace widget
- finish up / clean up widget button functionality

## Motivation and Context
This stems from a request for templated checkouts, which boils down to having find-and-replace functionality.  A naive find-and-replace on the json could touch file/field names, and is rather difficult to inspect.  

My intention is to make the future "fill template" step / tool wrap this functionality, possibly with some extra visualization or streamlining buttons.  A sample checkout could be run through this tool and and "filled" as is with this tool, I imagine

(In reality I implemented the simple find-and-replace helpers, tried making it more robust, and a week later ended up here.)

Creating a step that fills templates mid-checkout is a possible next step, though we'd have to consider expanding each template step on preparation and inserting it into the tree.  Inspection of the substeps of that filled template would be un-inspectable unless we expand in edit-mode, and in that case we're basically creating a bunch of filled template checkouts separately.

## How Has This Been Tested?
Some unit tests have been added, more need to be...

## Where Has This Been Documented?
This PR

## Screenshots 
image

image

image

## Pre-merge checklist
- [x] Code works interactively
- [x] Code contains descriptive docstrings, including context and API
- [x] New/changed functions and methods are covered in the test suite where possible
- [x] Code has been checked for threading issues (no blocking tasks in GUI thread)
- [x] Test suite passes locally
- [x] Test suite passes on GitHub Actions
- [x] Ran ``docs/pre-release-notes.sh`` and created a pre-release documentation page
- [x] Pre-release docs include context, functional descriptions, and contributors as appropriate

### CODE REVIEW

**Comment by tangkong** [comment]
This is basically feature complete as far as I care for.  There was a good amount of gui-ing here so I'm hoping for some adversarial clicking while I write comments / unit tests.

File: atef/ui/fill_template_page.ui
Comment (ZLLentz) [inline atef/ui/fill_template_page.ui:1]: Clickly clicky seemed to work as I used this and it seems very powerful in a way that I think will be well-appreciated by the users. A few thoughts:
- Subbing into a big structure can sometimes show a bit too much info. Is there any way we can truncate this so we only show the bit that changed?
- Is there any simple way we can automatically refresh when I add a new edit? I got confused for a moment at first before realizing that I needed to click the refresh button on the left.

image

Comment (ZLLentz) [reply atef/ui/fill_template_page.ui:1]: Oh, also: after saving the file, it seems like this whole template window goes blank. I think that will scare some people. I wonder if we should automatically close this and open the file that was just saved?

Comment (tangkong) [reply atef/ui/fill_template_page.ui:1]: > Subbing into a big structure can sometimes show a bit too much info. Is there any way we can truncate this so we only show the bit that changed?

I gave this my best effort, but I wonder if I can do any better.  Currently it shows the deepest dataclass, which I thought was needed to give the change any context at all.  But I suppose now that I have {object}.{attr} on the left I could just show the {attr} on the right

> Is there any simple way we can automatically refresh when I add a new edit? I got confused for a moment at first before realizing that I needed to click the refresh button on the left.

This is a perpetual issue I have.  I always try to hook more `editingFinished` signals, but find them to either not catch all the cases I'd want or trigger too often.  I'll see what I can do here

**Review by ZLLentz — APPROVED** [review]
I'm very positive on this. Pending Ken's review I think this is good enough to merge.

File: atef/widgets/config/window.py
Comment (ZLLentz) [reply atef/widgets/config/window.py:378]: this is really nifty

I wonder if anyone will get confused by the interface though- it wasn't obvious to me at first that this wasn't doing a find/replace in the same window, rather it is creating a whole new version that I could choose to edit and save seperately

File: atef/widgets/config/find_replace.py
Comment (tangkong) [reply atef/widgets/config/find_replace.py:505]: Ah I dropped my train of thought here.  `FindReplaceAction.apply` catches errors now, so I need that value to decide whether or not application was successful

File: atef/widgets/config/find_replace.py
Comment (tangkong) [reply atef/widgets/config/find_replace.py:956]: The caveat here is that the `TableWidgetWithAddRow` takes care of adding rows to the point where it never exposes those RowWidgets.

Adding a signal on this row that the table widget expects would force me to change the other instances of this table-row-thing to match, which I felt wasn't worth it at this juncture.  I totally agree though, this isn't quite how things should be done.

File: atef/ui/fill_template_page.ui
Comment (tangkong) [reply atef/ui/fill_template_page.ui:1]: I've made the previews show just the bit that changed (the last part of the path), as well as gave the user some notice that their file was saved and that they can open it.  

I'll give a bit more information in the details list if the user hasn't picked an edit as well

File: atef/widgets/config/find_replace.py
Comment (klauer) [inline atef/widgets/config/find_replace.py:814]: Hmm, it wasn't clear to me why the "edit details" weren't showing:
image
I think this should probably read "Click on edit in the left list to see details; if nothing shows here there is no match" or something to that effect

File: atef/widgets/config/find_replace.py
Comment (klauer) [inline atef/widgets/config/find_replace.py:876]: If I click 'apply all' after having changed a device in the template, I'd have expected to see the device/signal list above change:
image

File: atef/widgets/config/window.py
Comment (klauer) [reply atef/widgets/config/window.py:378]: Yeah, it is really cool but I also felt the same regarding the end user's expectations with the dialog.

"Apply All" seems like it should be replacing it in the file, but it's just doing the changes on an in-memory version until you click 'save as' (... right?)
Maybe the dialog should be something like:
> "Choose input file: "
> "Choose output file: "
> (Make edits in this section)
> Reload / Preview and Verify / Save

File: atef/widgets/config/find_replace.py
Comment (klauer) [inline atef/widgets/config/find_replace.py:567]: I think the biggest feature you could add is the ability to change (add/remove/edit) devices and signals directly in this table using the device/component selector dialog:
image

The find/replace stuff is very valuable, but I think the primary thing people will want to template is "test that device instead of this device"

File: atef/widgets/config/find_replace.py
Comment (klauer) [inline atef/widgets/config/find_replace.py:225]:
```suggestion
        # simple field paths don't have a final (__sth__, ?) segment
```

File: atef/widgets/config/find_replace.py
Comment (klauer) [inline atef/widgets/config/find_replace.py:100]: Checking the coverage - running test suite locally, it doesn't touch this line
(`coverage html && open htmlcov/index.html`)

GUI stuff aside, in my opinion we should try to aim for ~100% coverage on core library stuff that deals with our dataclasses. (Easier said than done of course!)

File: atef/widgets/config/find_replace.py
Comment (klauer) [inline atef/widgets/config/find_replace.py:251]:
```suggestion
    """Check to make sure changes are valid."""
```

File: atef/widgets/config/find_replace.py
Comment (klauer) [inline atef/widgets/config/find_replace.py:291]:
```python
ReplaceFunction = Callable[...something here...]
```
and
```suggestion
        replace_fn: Optional[ReplaceFunction] = None
```

Would recommend typing the "replace_fn" so you can tell if it's valid

File: atef/widgets/config/find_replace.py
Comment (klauer) [inline atef/widgets/config/find_replace.py:303]:
```suggestion
            The file to apply the find-replace action to, by default this applies to the current target of the action
```

Or something like that?

File: atef/widgets/config/find_replace.py
Comment (klauer) [inline atef/widgets/config/find_replace.py:418]: Abbreviated method name with star args/kwargs and no docstring... Sorry to be picky (it's genetic, so you're supposed to excuse it I think), but would you mind either expanding out the name or adding brief intent docstrings to these?

File: atef/widgets/config/find_replace.py
Comment (klauer) [inline atef/widgets/config/find_replace.py:713]: Shouldn't there be an implicit "verify_changes" call before it's saved?

File: atef/widgets/config/find_replace.py
Comment (klauer) [inline atef/widgets/config/find_replace.py:711]: Careful, `self.orig_file` can be `None` here per the static analyzer

File: atef/widgets/config/find_replace.py
Comment (klauer) [inline atef/widgets/config/find_replace.py:414]: This should probably `raise` and clear `original_json` 
At the very least `data` is unbound in this except clause

File: atef/widgets/config/find_replace.py
Comment (klauer) [inline atef/widgets/config/find_replace.py:590]: Probably can do better than `Any` here

File: atef/widgets/config/find_replace.py
Comment (klauer) [inline atef/widgets/config/find_replace.py:855]: `data` unused?

File: atef/widgets/config/find_replace.py
Comment (klauer) [inline atef/widgets/config/find_replace.py:774]: Dealing with `None` is annoying with respect to the static analyzer but I find it to be good practice - even if you're 100% sure this doesn't get called when `fp is None` - to guard for it anyway. A future refactor you may forget about it and then get confused why `basename(None)` is getting called

File: atef/widgets/utils.py
Comment (klauer) [inline atef/widgets/utils.py:260]: Aside: we should probably move to making an atef stylesheet at some point and move these things there (`atef.qss`, oh yeah!)

**Review by klauer — APPROVED** [review]
The functionality here is awesome! Nicely done. I think this will serve the end user rather well.

Some thoughts, suggestions, nitpicks, and musings as per usual. Nothing at a level that requires holding this PR up any longer. Sorry it took me so long to get around to reviewing!

File: atef/widgets/config/find_replace.py
Comment (tangkong) [reply atef/widgets/config/find_replace.py:590]: TIL `typing.TYPE_CHECKING` was a thing.  circular imports be gone!

File: atef/widgets/config/find_replace.py
Comment (tangkong) [reply atef/widgets/config/find_replace.py:855]: My `TableWidgetWithAddRow` expected its row widgets to be `DataWidget`s, all of which have a `data` kwarg.  I probably could have made a dataclass to represent an "Edit" but apparently past me didn't feel like it 🤷

File: atef/widgets/utils.py
Comment (tangkong) [reply atef/widgets/utils.py:260]: I wonder if that'd fix some of the weird osx rendering differences 🤔 A good idea, but I'm not volunteering right now... 👀

File: atef/ui/fill_template_page.ui
Comment (tangkong) [reply atef/ui/fill_template_page.ui:1]: I did think about this, but punted it never to look back.  I think it's honestly the ideal visualization, but involves quite a bit of work I think.  (Particularly because the tree does not often match the dataclass structure)

This is definitely worth a follow up issue methinks

File: atef/widgets/config/find_replace.py
Comment (tangkong) [reply atef/widgets/config/find_replace.py:567]: aaaah now that's a fancy idea.  I'll make a followup, because my vision of this quickly became very complicated

File: atef/widgets/config/find_replace.py
Comment (tangkong) [reply atef/widgets/config/find_replace.py:100]: I should do this more regularly in my code review as well as devlopment

File: atef/widgets/config/find_replace.py
Comment (tangkong) [reply atef/widgets/config/find_replace.py:418]: Don't be sorry, it's entirely your prerogative to to complain about this.  I should be less lazy.

File: atef/widgets/config/find_replace.py
Comment (tangkong) [reply atef/widgets/config/find_replace.py:95]: This would have probably made things easier.  I'll choose to pass on this, but toss it into an issue if later me decides I have extra cycles

File: atef/widgets/config/find_replace.py
Comment (tangkong) [reply atef/widgets/config/find_replace.py:405]: Good catch, this is left over from when I was editing the json directly.  Both here and in the FindReplaceWidget, the original json is never used.

Comment (tangkong) [reply atef/widgets/config/find_replace.py:405]: Also you're totally right we do have loaders. 🤦

File: atef/widgets/config/find_replace.py
Comment (tangkong) [reply atef/widgets/config/find_replace.py:713]: I agree with verifying as long as it doesn't prevent saving, since users might use this in an environment where they don't have access to all the necessary devices, making preparation fail.

Comment (klauer) [reply atef/widgets/config/find_replace.py:713]: Something like "Verification failed! Save anyway? [y/n]"

**Comment by tangkong** [comment]
I think I've addressed all the comments that fit as add-ons to this PR, and made issues for the rest.  I'd like to merge this and move on, unless there are any objections

**Comment by klauer** [comment]
Merge at will!

**Pull request merged**
